### PR TITLE
[BUGFIX] provide API to ensure the stream is  ended

### PR DESCRIPTION
**Pull request by stefanpenner**
related downstream PR: https://github.com/ember-cli/broccoli-concat/pull/44

- [x] need to sort out all versions of node (sorted out 0.10.x) – I believe it is related to: https://github.com/nodejs/node-v0.x-archive/issues/6118

### CODE REVIEW

**Comment by stefanpenner** [comment]
not quite correct, will fix it up in abit.

**Comment by stefanpenner** [comment]
@ef4 Im not terribly happy with the final implementation of this, but I would love another set of eyes.

File: lib/source-map.js [hunk 1]
```diff
@@ -362,22 +361,54 @@ SourceMap.prototype._scanMappings = function(srcMap, sourcesOffset, namesOffset,
...
+        reject(error);
+      });
+
+      stream.end();
+    });
+
+    var error, success;
+
+    try {
+      if (cb) {
```
Comment (kellyselden) [inline lib/source-map.js]: `if (typeof cb === 'function') {`?

File: lib/source-map.js [hunk 2]
```diff
@@ -362,22 +361,54 @@ SourceMap.prototype._scanMappings = function(srcMap, sourcesOffset, namesOffset,
...
+      mkdirp.sync(path.dirname(sourceMap.mapFile));
+      fs.writeFileSync(sourceMap.mapFile, JSON.stringify(sourceMap.content));
+      success = true;
+    } catch(e) {
+      success = false;
+      error = e;
+    } finally {
+      stream.on('close', function() {
+        if (success) {
+          resolve();
```
Comment (kellyselden) [inline lib/source-map.js]: Just a note, but if [`stream.on('finish'`](https://github.com/ef4/fast-sourcemap-concat/pull/19/files#diff-cbf7cad5c2d14ff5f8c1850747b7855bL374) above had any arguments, they would be lost here.

Comment (stefanpenner) [inline lib/source-map.js]: the docs do not appear to suggest there are any meaningful arguments.

File: lib/source-map.js [hunk 3]
```diff
@@ -362,22 +361,54 @@ SourceMap.prototype._scanMappings = function(srcMap, sourcesOffset, namesOffset,
...
+      });
+
+      stream.end();
+    });
+
+    var error, success;
+
+    try {
+      if (cb) {
+        cb.call(thisArg, sourceMap);
```
Comment (kellyselden) [inline lib/source-map.js]: Should the callback be after the `writeFileSync`?

Comment (stefanpenner) [inline lib/source-map.js]: no, that callback is what populates `this.content` that is written out in `writeFileSync`

File: lib/source-map.js (hunk 1 above)
Comment (stefanpenner) [inline lib/source-map.js]: Ya seems good, i suspect arity check and verification / assert of the callback will be good additions.

**Comment by ef4** [comment]
LGTM.

**Comment by ef4** [comment]
Published in 0.2.7.

**Pull request merged**